### PR TITLE
Fix template instance in template should be marked partially instantiated

### LINKED ISSUE #1864: Repeated elements in docs
There's something strange going on with the docs, where types from the standard library are documented repeatedly (with identical, empty content):

https://microsoft.github.io/typespec/standard-library/rest/reference/data-types#TypeSpec.Rest.Resource.ResourceCreateModel

image

https://github.com/microsoft/typespec/blob/main/docs/standard-library/rest/reference/data-types.md

image

This makes the already sparse standard library documentation extremely hard to read.

### PR DESCRIPTION
**Pull request by timotheeguerin**
fix #1864

Problem was given a template

```
model Foo<TFoo> {
  bar: Bar<TFoo>
}

model Bar<TBar> {
  t: TBar;
}
```

`Bar` would get partially instantiated inside of the `Foo` declaration but with using `TFoo` as argument which would cause the templateMapper to not be added and the decorators to not run(that part is expected)

Then in a semantic navigator you would get a reference to this partial type but not know it is partial.

Fix is to make sure the mapper is always added but is marked with `partial: true` if any of the template arguments are still template parameters.

image

### CODE REVIEW

**Comment by bterlson** [comment]
Looks good. Discussed with Tim offline the following points:

* isTemplateDeclaration should only return true for actual declarations, so no partial instantiations.
* isTemplateInstance should return true only for templates that are fully instantiated with non-template parameter types (i.e. are safe to emit).
* The middle two instances are, shall we say, "weird", and in general I suspect that the above two checks will be sufficient, but Tim also suggested another helper isTemplateInstantiable to cover the 2nd case in case we have a need to know whether a template type is waiting for more arguments in order to be a full fledged templateInstance.

**Review by markcowl — CHANGES_REQUESTED** [review]
Can you create a PR into the typespec-azure repo so we can have a look at any impacts ?  Otherwise, lgtm

File: packages/compiler/core/types.ts [hunk 1]
```diff
@@ -41,6 +41,15 @@ export interface BaseType {
   projectionSource?: Type;
   projectionBase?: Type;
   projector?: Projector;
+
+  /**
+   * Reflect if a type has been finished(Decorators have been called).
+   * There is multiple reasons a type might not be finished:
+   * - a template declaration will not
```
Comment (daviwil) [inline packages/compiler/core/types.ts:48]: Will not be finished?

**Review by daviwil — APPROVED** [review]

File: packages/compiler/core/types.ts (hunk 1 above)
Comment (timotheeguerin) [reply packages/compiler/core/types.ts:48]: yeah finished being when decorators are run. Decorators are not run on a template declaration

**Pull request merged**